### PR TITLE
Add example: GPT2 text generation with KerasNLP

### PR DESCRIPTION
**Pull request by chenmoneygithub**
This will be the companion tutorial with KerasNLP 0.5 release. Per our earlier discussion, this will be an example instead of developer guide.

### CODE REVIEW

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+and ~15G GPU RAM) since you will finetune the GPT-2 model. Running this tutorial on CPU
+runtime will take hours.
+"""
+
+"""
+## Install KerasNLP and Import Dependencies.
+"""
+
+"""shell
+!pip install -q -U git+https://github.com/keras-team/keras-nlp.git@master
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:34]: I will leave this open as a reminder. thanks!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+end = time.time()
+print("TOTAL TIME ELAPSED: ", end - start)
+
+"""
+## Into the Sampling Method
+
+In KerasNLP, we offer a few sampling method, e.g., contrastive search,
+Top-K search and beam search. By default our `GPT2CausalLM` uses contrastive
+search, but you can choose your own sampling method.
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:262]: Before our release we should possibly set the default to contrastive search, some manual checks are required though. Or we can do the switch as a release followup?

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+
+There are two ways to specify your custom sampler:
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:264]: thanks, done!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+gpt2_lm.compile(sampler="top_k")
+
+start = time.time()
+
+output = gpt2_lm.generate("I like basketball", max_length=200)
+print("\nGPT-2 output:")
+print(output)
+
+end = time.time()
+print("TOTAL TIME ELAPSED: ", end - start)
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:281]: removed!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+
+"""
+For more details on KerasNLP `Sampler` class, you can check the code
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:297]: sg! leave this open as a reminder.

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+!pip install -q -U git+https://github.com/keras-team/keras-nlp.git@master
+"""
+
+import keras_nlp
+import tensorflow as tf
+from tensorflow import keras
+import time
+
+"""
+# Introduction to Generative Large Language Models (LLMs)
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:43]: simplified!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+
+import tensorflow_datasets as tfds
+
+reddit_ds = tfds.load("reddit_tifu", split="train", as_supervised=True)
+
+"""
+Let's take a look inside sample data from the reddit TensorFlow Dataset. There
+are two features:
+
+- __document__: text of the post
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:194]: good call, done

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+
+This step takes quite a bit of GPU memory and a long time if we were to train
+it all the way to a fully trained state. Here we just use part of the dataset
+for demo purposes.
+"""
+
+train_ds = train_ds.take(500)
+num_epochs = 1
+
+lr = tf.keras.optimizers.schedules.PolynomialDecay(
```
[the remark this answers is not in the document]
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:229]: i would like to keep it simple for now.. putting a custom LR schedule is a bit distracting IMO. Gabriel is working on adding warmup support to core Keras, I will check with him about the timeline.

[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:216]: sg!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+
+"""
+## Install KerasNLP and Import Dependencies.
+"""
+
+"""shell
+!pip install -q -U git+https://github.com/keras-team/keras-nlp.git@master
+"""
+
+import keras_nlp
```
Comment (fchollet) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:36]: Also import keras

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+
+This step takes quite a bit of GPU memory and a long time if we were to train
+it all the way to a fully trained state. Here we just use part of the dataset
+for demo purposes.
+"""
+
+train_ds = train_ds.take(500)
+num_epochs = 1
+
+lr = tf.keras.optimizers.schedules.PolynomialDecay(
```
Comment (fchollet) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:216]: Replace all `tf.keras` with `keras`

**Review by fchollet — COMMENTED** [review]
Great stuff!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+pedagogical discussion on language models, you can refer to the
+[Stanford CS324 LLM class](https://stanford-cs324.github.io/winter2022/lectures/introduction/).
+"""
+
+"""
+## Introduction to KerasNLP
+
+Large Language Models are complex to build and expensive to train from scratch.
+Luckily there are pretrained LLMs available for use right away. One toolkit that
+offers state-of-the-art pretrained models for free is [KerasNLP](https://keras.io/keras_nlp/).
```
Comment (mattdangerw) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:66]: "for free" kinda makes us feel like too much of a salesman here for my taste. I would probably just say

KerasNLP provides a large number of pre-trained checkpoints that allow you to experiment with SOTA models without needing to train them yourself.

or something like that

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+
+KerasNLP provides a number of pre-trained models, such as [Google
+Bert](https://ai.googleblog.com/2018/11/open-sourcing-bert-state-of-art-pre.html)
+and [GPT-2](https://openai.com/research/better-language-models). You can see
+the list of models available in the [KerasNLP repository](https://github.com/keras-team/keras-nlp/tree/master/keras_nlp/models).
+
+It's very easy to load the GPT-2 model as you can see below:
+"""
+
+# To speed up, we use preprocessor of length 128 instead of full length 1024.
```
Comment (mattdangerw) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:93]: To speed up training and generation, ...

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+preprocessor = keras_nlp.models.GPT2CausalLMPreprocessor.from_preset(
+    "gpt2_base_en",
+    sequence_length=128,
+    add_end_token=True,
```
Comment (mattdangerw) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:97]: we no longer need this i think

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+output = gpt2_lm.generate("That Italian restaurant is", max_length=200)
+print("\nGPT-2 output:")
+print(output)
+
+end = time.time()
+print(f"TOTAL TIME ELAPSED: {end - start:.2f}s")
+
+"""
+Notice how much faster the second call is. This is because the computational
+graph is compiled in the 1st run and re-used in the 2nd behind the scene.
```
Comment (mattdangerw) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:133]: behind the scene -> behind the scenes

Comment (mattdangerw) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:133]: also we might want to name drop XLA in the guide somewhere, with a link to https://www.tensorflow.org/xla

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+
+The quality of the generated text looks OK, but we can improved it via
+finetuning.
+"""
+
+"""
+## More on the GPT-2 model from KerasNLP
+
+While it may be sufficient to move on to the next step of finetuning the
```
Comment (mattdangerw) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:142]: This is a slightly weird intro paragraph flow wise. Maybe just say.

Next up, we will actually fine-tune the model to update it's parameters, but before we do, let's take a look at the full set of tools we have to for working with for GPT2.

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+it all the way to a fully trained state. Here we just use part of the dataset
+for demo purposes.
+"""
+
+train_ds = train_ds.take(500)
+num_epochs = 1
+
+learning_rate = keras.optimizers.schedules.PolynomialDecay(
+    5e-5,
+    decay_steps=train_ds.cardinality() * num_epochs,
```
Comment (mattdangerw) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:219]: should we show `power=1.0` explicitly, and mention this is a linear decay? i would find this confusing as a user

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:252]: i think we should just stick with top-k for now as that's what we just pushed out with

also top-k is fast, so not a bad choice!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+sampler:
+
+- Use a string identifier, such as "top_k", you are using the default
+configuration via this way.
+- Pass a `keras_nlp.samplers.Sampler` instance, you can use custom configuration
+via this way.
+"""
+
+# Use a string identifier.
+gpt2_lm.compile(sampler="top_k")
```
Comment (mattdangerw) [inline examples/nlp/gpt2_text_generation_with_kerasnlp.py:264]: maybe use some samplers that will actually show noticeable difference here, might be cool for a reader

`sampler="greedy"`
`sampler="random"`
`sampler=keras_nlp.samplers.TopK(k=10)`

And make some comments about how greedy will tend to repeat itself, random sampling can tend to choose weird outputs, and top-k will be somewhere in the middle?

**Review by mattdangerw — APPROVED** [review]
Left a last few rounds of comments. Looking good!

File: examples/nlp/gpt2_text_generation_with_kerasnlp.py (hunk 16 above)
Comment (chenmoneygithub) [reply examples/nlp/gpt2_text_generation_with_kerasnlp.py:219]: Added a comment.

**Review by fchollet — APPROVED** [review]
LGTM
